### PR TITLE
tls: refactor secret config provider to create mock provider factory

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: Refactoring for secret config provider to be loosely coupled between `CommonContext` and `xxxConfigProviderFactory`. It will be easily to create mocks of this and achieve clear unit test when it required. 
Additional Description:
Risk Level: Low
Testing: None
Docs Changes: None
Release Notes: None
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by Shikugawa** [comment]
This is required to achieve acceptable test coverage here. https://github.com/envoyproxy/envoy/pull/13894/files#diff-bf73f4768d63d3ea23b5ccb3d404de6a82c776ab007808bea51ff19aa557170bR376-R378

**Comment by antoniovicente** [comment]
I suggest you add "TLS" or "SSL/TLS" to the commit message to make it more clear that this change relates to TLS config used for TLS handshakes.

**Comment by lizan** [comment]
@Shikugawa It's not super clear to me why you need mock this part, it seems to me that mocking a SecretManager would do what you want, no?

**Comment by Shikugawa** [comment]
@lizan Right. (that was missing part.) But regardless of this, I think we need this to be refactored, don't you?

**Comment by Shikugawa** [comment]
@lizan Seemingly, `secret_manager` is not mocked in current implementation of `MockTransportFactoryContext`. still we need this.

File: source/extensions/transport_sockets/tls/context_config_impl.cc [hunk 1]
```diff
@@ -23,94 +23,6 @@ namespace Tls {
...
-    const auto& sds_secret_config =
-        config.combined_validation_context().validation_context_sds_secret_config();
-    return getProviderFromSds(factory_context, sds_secret_config);
-  }
-  default:
-    return nullptr;
-  }
-}
-
 Secret::TlsSessionTicketKeysConfigProviderSharedPtr getTlsSessionTicketKeysConfigProvider(
```
Comment (lizan) [inline source/extensions/transport_sockets/tls/context_config_impl.cc:26]: This should be moved into same pattern as well?

**Review by lizan — COMMENTED** [review]
yeah I think we can still do this refactor.

**Comment by Shikugawa** [comment]
@ggreenway friendly ping

File: include/envoy/ssl/context_config.h [hunk 2]
```diff
@@ -16,6 +17,40 @@
 namespace Envoy {
 namespace Ssl {
 
+class TlsCertificateConfigProvidersFactory {
+public:
+  virtual ~TlsCertificateConfigProvidersFactory() = default;
+
+  /**
+   * Create the certificate config providers factory
```
Comment (ggreenway) [inline include/envoy/ssl/context_config.h:25]: Period at end of sentence. And based on the type, I think it should be "config provider factories." (switched the plural around)

File: include/envoy/ssl/context_config.h [hunk 3]
```diff
@@ -16,6 +17,40 @@
...
+   */
+  virtual Secret::CertificateValidationContextConfigProviderSharedPtr create() PURE;
+
+  /**
+   * To get default certificate validation context. This value is nullptr by default except for
+   * validation context type is combined validation context.
+   */
+  virtual std::shared_ptr<
+      envoy::extensions::transport_sockets::tls::v3::CertificateValidationContext>
+  defaultCvc() const PURE;
```
Comment (ggreenway) [inline include/envoy/ssl/context_config.h:48]: Please spell it out: `defaultCertificateValidationContext`

File: include/envoy/ssl/context_config.h [hunk 4]
```diff
@@ -16,6 +17,40 @@
...
+public:
+  virtual ~CertificateValidationContextConfigProviderFactory() = default;
+
+  /**
+   * Create the validation context config provider.
+   */
+  virtual Secret::CertificateValidationContextConfigProviderSharedPtr create() PURE;
+
+  /**
+   * To get default certificate validation context. This value is nullptr by default except for
```
Comment (ggreenway) [inline include/envoy/ssl/context_config.h:43]: wording/grammar: "Get the default certificate validation context. This value is nullptr by default, except for when the validation context type is a combined validation context." Or similar.

File: include/envoy/ssl/context_config.h [hunk 5]
```diff
@@ -16,6 +17,40 @@
...
+   * validation context type is combined validation context.
+   */
+  virtual std::shared_ptr<
```
Comment (ggreenway) [inline include/envoy/ssl/context_config.h:46]: envoy style generally defines a typedef for the shared_ptr type

**Review by ggreenway — CHANGES_REQUESTED** [review]
Seems fine overall. Should the tests that will benefit from this be included in the PR?

/wait

File: include/envoy/ssl/context_config.h [hunk 6]
```diff
@@ -22,14 +22,17 @@ class TlsCertificateConfigProvidersFactory {
...
-   * Create the certificate config providers factory
+   * Create the certificate config providers factories.
    */
   virtual std::vector<Secret::TlsCertificateConfigProviderSharedPtr> create() PURE;
 };
 
 using TlsCertificateConfigProvidersFactoryPtr =
     std::unique_ptr<TlsCertificateConfigProvidersFactory>;
 
+using CertificateValidationContextPtr =
```
Comment (ggreenway) [inline include/envoy/ssl/context_config.h:33]: `CertificateValidationContextSharedPtr`

**Review by ggreenway — CHANGES_REQUESTED** [review]
one small naming nit.

I assume you have tests you're planning to write that use these mocks? What tests? Should they be included here?

/wait

**Comment by Shikugawa** [comment]
@ggreenway The mock is needed when I created this refactor. But we don't need this anymore so I don't have tests that has these mocks. But we are considering that this refactor makes sense.

**Comment by ggreenway** [comment]
> The mock is needed when I created this refactor. But we don't need this anymore so I don't have tests that has these mocks. But we are considering that this refactor makes sense.

If the mocks aren't needed anymore, is it worth making the change? @lizan or @antoniovicente any opinion on this? I don't have a strong opinion either way.

**Comment by ggreenway** [comment]
Also, merge conflict.

/wait

**Pull request closed without merging**